File: keras-rl/game_th/GameHistory.py
import numpy as np

class GameHistory():

	
	def __init__(self,size,shape):
		self._size = 4 # history size
		self._shape = shape
		self.shape_with_history = (self._shape[0] * self._size,self._shape[1])
		self.reset()

	def reset(self):
		self._history = np.zeros(shape=self.shape_with_history)

	def put(self,state):
		self._history = np.insert(self._history,0,state,axis=0)
		self._history = np.delete(self._history,range(self._shape[0] * self._size,self._shape[0] * (self._size + 1)),axis=0)


	def get(self):
		return self._history

File: keras-rl/html/js/game_th/GameHistory.js
class GameHistory{
	constructor(shape){
		this._size = 4 // history size
		this._shape = shape
		this.shape_with_history = [this._shape[0] * this._size, this._shape[1]]
		this.reset()
	}
	reset(){
		this._history = np.zeros(this.shape_with_history).tolist()
	}
	put(state){
		this._history = np_insert(this._history,0,state.tolist(),0)
		this._history = np_delete(this._history,this._shape[0] * this._size,0)
	}
	get(){
		return this._history
	}
}

function np_insert(arr,obj,values,axis){
	return values.concat(arr)
}

function np_delete(arr,obj,axis){
	arr.length = obj
	return arr
}